gammapy/maps/sparse.py

```diff
@@ -0,0 +1,169 @@
+    for indexing and slicing logic.
+    """
+
+    def __init__(self, shape, idx=None, data=None, dtype=float, fill_value=0.0):
+
+        self._shape = tuple(shape)
+        self._fill_value = fill_value
+        if idx is not None:
+            self._idx = idx
+            self._data = data
+        else:
+            self._idx = np.zeros(0, dtype=int)
+            self._data = np.zeros(0, dtype=dtype)
+
+    @property
+    def size(self):
+        """Return current number of elements."""
+        return len(self._data)
+
+    @property
+    def data(self):
+        """Return the sparsified data array."""
+        return self._data
+
+    @property
+    def idx(self):
+        """Return flattened index vector."""
+        return self._idx
+
+    @property
+    def shape(self):
+        """Return the array shape."""
+        return self._shape
+
+    @property
+    def ndim(self):
+        """Return the array dimension."""
+        return len(self._shape)
+
+    @classmethod
+    def from_array(cls, data, min_value=0):
+        """Create from a numpy array.
+
+        Parameters
+        ----------
+        data : `numpy.ndarray`
+            Input data array.
+
+        min_value : float
+            Threshold for sparsifying the data vector.
+        """
+        shape = data.shape
+        idx = np.where(data > min_value)
+        idx = np.ravel_multi_index(idx, shape)
+        data = data[data > min_value]
+        return cls(shape, idx, data)
+
+    def _to_flat_index(self, idx_in):
+        """Convert index tuple to a flattened index."""
+        idx_in = tuple([np.array(z, ndmin=1, copy=False) for z in idx_in])
+        msk = np.all(
+            np.stack([t < n for t, n in zip(idx_in, self.shape)]), axis=0)
+        idx = np.ravel_multi_index(tuple([t[msk] for t in idx_in]),
+                                   self.shape, mode='wrap')
+
+        return idx, msk
+
+    def set(self, idx_in, vals, fill=False):
+        """Set array values at indices ``idx_in``."""
+
+        o = np.broadcast_arrays(vals, *idx_in)
+        vals = np.ravel(o[0])
+
+        # TODO: Determine whether new vs. existing indices are being
+        # addressed, in the latter case we only need to update data
+        # array
+
+        vals = np.array(vals, ndmin=1)
+        idx_flat_in, msk_in = self._to_flat_index(idx_in)
+        idx, data = merge_sparse_arrays(idx_flat_in, vals,
+                                        self.idx, self.data,
+                                        fill=fill)
+
+        self._idx = idx
+        self._data = data
+        #idx, msk = find_in_array(idx_flat_in, self.idx)
+        #self._data[idx[msk]] = vals[msk]
+
+    def get(self, idx_in):
+        """Get array values at indices ``idx_in``."""
+
+        shape_out = idx_in[0].shape
+        size_out = idx_in[0].size
+        idx_flat_in, msk_in = self._to_flat_index(idx_in)
+
+        # Get broadcasted shape?
+
+        idx, msk = find_in_array(idx_flat_in, self.idx)
```

Comment:
@woodmd - Concerning the buffer dtype mismatch fails on Windows:
https://ci.appveyor.com/project/cdeil/gammapy/build/1.0.2754/job/vop5ndc90lbhdlsd#L1572

I think the solution is to use Numpy dtypes `int64` and `float64` in both the calling Python code and the receiving Cython code. I.e. change all the Cython code to have fully explicity `int64` and `float64` types and then on the Python side, add a `data = np.asanyarray(data, dtype=np.int64)` at the top of the function, to make sure you're calling the Cython function with the correct number type.

See e.g.
https://stackoverflow.com/questions/33090711/cython-valueerror-dtype-mismatch-expected-a-int-but-got-long-long
or several examples in other packages like https://github.com/astropy/astropy-healpix/tree/master/astropy_healpix

@woodmd - Can you try? Or should I try to edit the types and make a commit here?
(I'm also not experienced with Cython, and don't have access to a Windows box, so it might be a bit of a trial and error game).

Reply:
Your proposed solution sounds good to me.  I'll give it a try fixing this myself.